csrc/preseg_passes/decompose_reshardings.cpp

```diff
@@ -162,118 +218,68 @@ void insertReshardingSetsAfter(Fusion* fusion) {
-      // WAR (https://github.com/NVIDIA/Fuser/pull/5205#issuecomment-3324173092)
-      // this pass might incorrectly overwrite output's DIDx parallelization,
-      // which should be avoided. This case shows up if the fusion lowers to a
-      // p2p ring pipeline and the option "reshard_after" is enabled. The proper
-      // fix should be to make parallelizeAllLike smarter and skip the
-      // propagation over a DID parallelization type. We leave this as is as a
-      // temporary solution, as the bug is catched by an assertion.
-      scheduler_utils::parallelizeAllLike(
-          new_output, {output}, {ParallelType::Stream});
-    }
-  }
-}
-
-// Canonicalizes tv's loop domain for simplicity and working around schedulers'
-// limitations. Many schedulers panic when seeing the input fusion segment
-// contains non-DID loop splits. For example, an rFactor tensor may look like
-// the following:
-//
-//                            r{k}
-//                            /  \.
-// [i{m}         i{n}    iDIDx{d}  r{k/d}]
-//               /  \.
-//            i{d} i{n/d}
-//
-// The split of i{n} is unnecessary because i{d} and i{n/d} are both
-// ParallelType::Serial. This function replaces the two with i{n} in the loop
-// domain.
-void canonicalizeLoopDomain(TensorView* tv) {
-  LinkedHashMap<IterDomain*, std::monostate> loop;
-  for (IterDomain* id : tv->getLoopDomain()) {
-    loop.pushBack(id, std::monostate());
-  }
-
-  for (Expr* transform :
-       DependencyCheck::getAllExprsBetween(
-           {tv->getLogicalDomain().begin(), tv->getLogicalDomain().end()},
-           {tv->getLoopDomain().begin(), tv->getLoopDomain().end()}) |
-           std::views::reverse) {
-    auto* split = dynamic_cast<Split*>(transform);
-    NVF_ERROR(
-        split != nullptr,
-        "Only splits are expected so far, but found: ",
-        transform);
-
-    if (split->outer()->isParallelized() || split->inner()->isParallelized()) {
+    if (resharding_input == nullptr) {
       continue;
     }
 
-    if (!loop.contains(split->outer()) || !loop.contains(split->inner())) {
-      continue;
-    }
-
-    loop.erase(split->outer());
-    const auto inner_i = loop.erase(split->inner()).second;
-    // `inner_i` is picked arbitrarily as the insertion point. Given `in`,
-    // `outer` and `inner` are all serial, `in`'s position in the loop domain
-    // doesn't matter.
-    loop.insert(inner_i, split->in(), std::monostate());
+    // Insert resharding set after the expr and update
+    // output of expr to match input's sharding.
+    TensorView* new_output = set(output);
+    ir_utils::replaceValInAllExprInputsAndFusionOutputs(output, new_output);
+    // Update shardings new_output takes output's sharding,
+    // output takes input's sharding
+    shardLoopLike(
+        /*ref=*/output,
+        /*target=*/new_output,
+        deviceAndStreamParallelTypes(),
+        PropagateDirection::kForward);
+
+    unshard(output);
```

Comment:
Can you code-comment why this is needed? I guess it's because shardLoopLike acts conservatively and gives up when a parallel type to propagate already exists in `output`.